plugins/roundedwindow/roundedwindow.cpp

```diff
@@ -172,9 +172,7 @@ static std::unique_ptr<KWin::GLShader> getShader()
     stream.flush();
 
     auto shader = KWin::ShaderManager::instance()->generateCustomShader(traits, QByteArray(), source);
-    //shaders.insert(direction, shader);
-    std::unique_ptr<KWin::GLShader> p(shader);
-    return p;
+    return shader;
```

Comment:
As the API of `generateCustomShader` is not consistent in different KWin versions. A better solution here is to check whether the `shader` ptr is a smart ptr or not.

You can refer to this example:

```cpp
#include <iostream>
#include <memory>
#include <type_traits>

// 通用模板，假定不是智能指针
template<typename T>
struct is_smart_pointer : std::false_type {};

// std::shared_ptr特化版本
template<typename T>
struct is_smart_pointer<std::shared_ptr<T>> : std::true_type {};

// std::unique_ptr特化版本
template<typename T>
struct is_smart_pointer<std::unique_ptr<T>> : std::true_type {};

// std::weak_ptr特化版本
template<typename T>
struct is_smart_pointer<std::weak_ptr<T>> : std::true_type {};

int main() {
    std::shared_ptr<int> sp = std::make_shared<int>(10);
    std::unique_ptr<int> up(new int(10));
    std::weak_ptr<int> wp(sp);
    int* p = new int(10);

    // 测试
    std::cout << "sp is smart pointer: " << is_smart_pointer<decltype(sp)>::value << std::endl;
    std::cout << "up is smart pointer: " << is_smart_pointer<decltype(up)>::value << std::endl;
    std::cout << "wp is smart pointer: " << is_smart_pointer<decltype(wp)>::value << std::endl;
    std::cout << "p is smart pointer: " << is_smart_pointer<decltype(p)>::value << std::endl;

    delete p; // 记得释放p指向的内存
    return 0;
}
```

Reply:
One more thing needs checking is that the convert to smart_ptr is somehow dangerous as it takes control of the ptr. It will delete it once on one references to the converted smart ptr. Make sure the original ptr is not used elsewhere.